[Yaml] Add --exclude and negatable --parse-tags option to lint:yaml command

christingruber opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | None
| License       | MIT
| Doc PR        | symfony/symfony-docs#14780

Added a new `--exclude` option to exclude one or more specific files from multiple file list:

```lint:yaml dirname --exclude=/dirname/foo.yml --exclude=/dirname/bar.yml``` 
 
Allow negatable for the parse tags option with `--no-parse-tags` and increase the `symfony/console` dependency to version 5.3

christingruber commented on src/Symfony/Component/Yaml/Tests/Command/LintCommandTest.php:
```diff
@@ -142,6 +142,35 @@ public function testCustomTagsError()
+    {
+        $tester = $this->createCommandTester();
+
+        copy(__DIR__.'/../Fixtures/.yamllint.yml', './.yamllint.yml');
```
Find a better way, fails with tests...

jderusse commented on src/Symfony/Component/Yaml/Command/LintCommand.php:
```diff
@@ -98,12 +110,25 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $this->displayCorrectFiles = $output->isVerbose();
         $flags = $input->getOption('parse-tags') ? Yaml::PARSE_CUSTOM_TAGS : 0;
 
+        if ([$this->lintConfig] === $filenames) {
```
hmm.. this provides an argument to load a config file, but prevent using any files.

What's about adding a new options `-c | --config` and checking if a `.yamllint.yaml` file exist in the project DIR

Or checking if the $filenames contains a single file and this file is a valid yamlint file

christingruber replied:
> What's about adding a new options `-c | --config`...

I think a new option is a good idea. Maybe with an optional value for a custom config file name `--config=[yamllint.yml]`.

What do you think?

xabbuh replied:
I would not provide a default config file, but always require to pass the config filename when this otpion is used.

xabbuh commented on src/Symfony/Component/Yaml/Command/LintCommand.php:
```diff
@@ -84,6 +95,7 @@ protected function execute(InputInterface $input, OutputInterface $output)
     {
         $io = new SymfonyStyle($input, $output);
         $filenames = (array) $input->getArgument('filename');
+        $excludes = explode(' ', $input->getOption('exclude'));
```
We can configure the option instead to accept multiple arguments using `InputOption::VALUE_IS_ARRAY`.

christingruber replied:
Changed :+1:

OskarStark commented on src/Symfony/Component/Yaml/Tests/Command/LintCommandTest.php:
```diff
@@ -142,6 +142,44 @@ public function testCustomTagsError()
+
+        $this->assertSame(0, $ret, 'lint:yaml exits with code 0 in case of success');
+        $this->assertStringContainsString('All 1 YAML files contain valid syntax.', trim($tester->getDisplay()));
+        unlink('.yamllint.yml');
```
We should use a tearDown() method to ensure the file is deleted after all tests, if one failed. Not sure how symfony handles tests like this

christingruber replied:
Done, but not final. I will move the path to `sys_get_temp_dir()` with the `--config option`.

christingruber replied:
There is now the `--config|-c` option without default value. I have also changed the tests, which will use the temp dir for the dummy lint config now.

BTW I have also added a shortcut `-e` for the exclude option.

xabbuh commented on src/Symfony/Component/Yaml/.yamllint.yml.dist:
```diff
@@ -0,0 +1,8 @@
+yaml-lint:
```
This file is no longer needed, is it?

christingruber replied:
Yes I think we can drop the file, an example config is documented.

xabbuh reviewed (commented):
I think we should also add a test that handles a call of the command with a config file being passed as well as other options.

christingruber commented:
> I think we should also add a test that handles a call of the command with a config file being passed as well as other options.

I've added a simple mixed test, ~~settings from yaml config will have priority in this case~~

christingruber commented:
I have changed the load order to allow overrides by console args...

christingruber commented:
A short overview for the last changes:

I have added some tests which handles mixed arguments from config and options from console. I have also changed the load order to allow overrides by console options. With that it is possible to override the values ​​provided from lint config by console options. @xabbuh

xabbuh commented on src/Symfony/Component/Yaml/Command/LintCommand.php:
```diff
@@ -84,7 +94,31 @@ protected function execute(InputInterface $input, OutputInterface $output)
+                $filenames = $lintConfig['includes'];
+            }
+            if (!\count($excludes) && isset($lintConfig['excludes'])) {
+                $excludes = $lintConfig['excludes'];
```
I think I would merge excludes from config file and CLI options.

christingruber replied:
OK changed.

xabbuh commented on src/Symfony/Component/Yaml/Command/LintCommand.php:
```diff
@@ -84,7 +94,31 @@ protected function execute(InputInterface $input, OutputInterface $output)
+                $flags = $lintConfig['parse-tags'];
+            }
+            if (!\count($filenames) && isset($lintConfig['includes'])) {
+                $filenames = $lintConfig['includes'];
```
I think I would merge includes from config file and CLI options.

christingruber replied:
OK changed.

xabbuh commented on src/Symfony/Component/Yaml/Command/LintCommand.php:
```diff
@@ -84,7 +94,31 @@ protected function execute(InputInterface $input, OutputInterface $output)
+            if (null === $this->format && isset($lintConfig['format'])) {
+                $this->format = $lintConfig['format'];
+            }
+            if (false === $flags && isset($lintConfig['parse-tags'])) {
```
We should probably turn the `--parse-tags` option into a negatable option so that one can override the config using either `--parse-tags` or `--no-parse-tags`.

christingruber replied:
The negatable option is only avail with symfony/console 5.3, right?

stof replied:
which is OK as you add this in symfony/yaml 5.3 too. You can bump the min symfony/console version.

christingruber replied:
I'm not sure with the negatable: `--parse-tags` give us `true` and `--no-parse-tags` give us `false`, but the default "without the option" is also `false`. So, currently I can not detect if the user give us the option or not and the console option wins always...

I think we have to set a default `null` for this option to handle the override, but this brings a BC... Am I right?

xabbuh replied:
Setting the default value to `null` should be fine. If I don't miss anything, that shouldn't break anything.

christingruber replied:
OK changed.

xabbuh commented on src/Symfony/Component/Yaml/Command/LintCommand.php:
```diff
@@ -113,7 +148,9 @@ protected function execute(InputInterface $input, OutputInterface $output)
 
             foreach ($this->getFiles($filename) as $file) {
-                $filesInfo[] = $this->validate(file_get_contents($file), $flags, $file);
+                if (!\in_array($file, $excludes)) {
```
please pass `true` as the third argument to `in_array()`

christingruber replied:
:+1:  changed.

xabbuh commented on src/Symfony/Component/Yaml/CHANGELOG.md:
````diff
@@ -6,6 +6,20 @@ CHANGELOG
+ * Added support for optional YAML lint config from file `lint:yaml --config=.yamllint.yml`.
+
+   ```yml
+   yaml-lint:
````
I am not really sure that we need this top-level entry. IMO this would work as well:

```yaml
format: txt
includes:
  - src/Resources/config/foo.yml
  - src/Resources/config/bar.yml
  - src/Resources/config/path/
excludes:
  - src/Resources/config/path/exclude.yml
```

christingruber replied:
Yes it works fine without the top level entry. The only technical reason for the tag is to check if the file has a "valid" lint config:

```php
if (!isset($result['yaml-lint'])) {
    throw new RuntimeException(sprintf('Invalid YAML lint config "%s".', $config));
}
```

And with the tag it is possible to mix the config with other configurations in the same file, not sure if someone need this but it is possible. 

Last but not least, the tag could help other developers to detect the reason for the config file...

For me both is OK. What do you think?

xabbuh replied:
I think that it will be easier to write a config from scratch if we ditch the top-level `yaml-lint` entry. I would then do a naiv validation which will do a check like if `array_diff(array_keys($config), ['format', 'includes', 'excludes', 'parse-tags'])` returns the empty array.

christingruber replied:
OK so, I have dropped the top level entry and changed the tests.

javiereguiluz commented:
I'm divided about this proposal. I love the new options that you added to the command but I'm not sure about providing support for a new config file specifically to run this command.

Just asking: if we keep all the new command options, could a makefile task be an alternative to this config file? You would add the values of all the command options to create a very long command ... but since it's a Makefile, you don't have to actually type or remember the long command, only a short alias. Maybe it's not possible, I don't know.

As a poor version of the Makefile, a simple `lint.sh` or `lint.bat` script in the `bin/` dir of the project could work too to avoid passing all the long option values. Cheers!

christingruber commented:
@javiereguiluz Thanks for your feedback.

No problem I think we can hold the `--exclude` option and drop the config part. If so, I can refactor and rebase the branch in the next days. OK?

javiereguiluz commented:
@christingruber thanks ... but maybe we should wait for other comments from @symfony/mergers first. The reason is that in my previous comment I only gave my opinion ... and a single person opinion is not important. Better see what's the general consensus about what to do here. Thanks!

Nyholm reviewed (changes requested):
Wow, what a great PR. Well done. I like that you added tests and a good description in the changelog. I also like the implementation. 

I do agree with @javiereguiluz, Let's not add a config file. If we would go the config file route, we must also support different formats and it would make sense to allow other commands to have config files too. It will be a large feature and Im not sure about that route. I think the workarounds mentioned before is better.

I would be happy to see this PR with only the `exclude` feature.

chalasr commented:
Thanks for your work on this.

> I would be happy to see this PR with only the exclude feature.

Same for me :)

OskarStark commented:
Hi @christingruber 👋 

Are you open to finish this PR for `5.4` 🎯 ?

Cheers

christingruber commented:
@OskarStark All or only the exclude option?

OskarStark commented:
Only the exclude please

christingruber committed:
[Yaml] Add an --exclude option to lint:yaml command

chalasr commented on src/Symfony/Component/Yaml/Tests/Command/LintCommandTest.php:
```diff
@@ -166,7 +176,7 @@ protected function createCommandTester(): CommandTester
         $application = new Application();
         $application->add(new LintCommand());
-        $command = $application->find('lint:yaml');
+        $command = $application->find(LintCommand::getDefaultName());
```
unrelated change should be reverted

Nyholm commented on src/Symfony/Component/Yaml/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
+5.4
+---
+
+* Added new `lint:yaml dirname --exclude=/dirname/foo.yml --exclude=/dirname/bar.yml` option to exclude one or more specific files from multiple file list.
```
Add new….

Nyholm commented on src/Symfony/Component/Yaml/Tests/Command/LintCommandTest.php:
```diff
@@ -142,6 +142,16 @@ public function testCustomTagsError()
+    {
+        $tester = $this->createCommandTester();
+        $filename = $this->createFile('foo: bar');
+
```
Let’s create and other file that we do parse.

Nyholm commented on src/Symfony/Component/Yaml/composer.json:
```diff
@@ -23,10 +23,10 @@
     },
     "conflict": {
-        "symfony/console": "<4.4"
+        "symfony/console": "<5.4"
```
Can you tell me why we need to update composer.json? 

Is it because the Negatable? Is that related to the exclude feature?

christingruber replied:
It is for the negatable `InputOption::VALUE_NEGATABLE` option.

chalasr replied:
Using a negatable option was suggested here https://github.com/symfony/symfony/pull/39641#discussion_r556372547

Nyholm reviewed (changes requested):
Great. Thank you for this PR. 

I added some questions. Also, can you make sure the PR description is up to date? It will make it easier to understand this feature after the PR is merged.

chalasr reviewed (approved):
I

christingruber commented:
> I added some questions. Also, can you make sure the PR description is up to date?...

Hi @Nyholm,

thanks for your Feedback, I updated the MR description / changelog and the composer deps. (See my further comments)

Nyholm reviewed (approved):
Thank you. 

I like this feature and the implementation. I also appreciate that you updated the PR description and that you prepared a doc PR.  

Well done.

OskarStark commented on src/Symfony/Component/Yaml/CHANGELOG.md:
```diff
@@ -1,6 +1,14 @@
+ * Add new `lint:yaml dirname --exclude=/dirname/foo.yml --exclude=/dirname/bar.yml`
+   option to exclude one or more specific files from multiple file list
+ * Allow negatable for the parse tags option with `--no-parse-tags` and increase
+   the `symfony/console` dependency to version 5.3
```
Not sure we need to mention the increase

fabpot replied:
Indeed

christingruber replied:
👍🏻  removed.

christingruber replied:
OK. Removed. (Give me a short hint if it's the other way around)

OskarStark reviewed (approved):
One minor

OskarStark commented on src/Symfony/Component/Yaml/CHANGELOG.md:
```diff
@@ -1,6 +1,14 @@
+5.4
+---
+
+ * Add new `lint:yaml dirname --exclude=/dirname/foo.yml --exclude=/dirname/bar.yml`
```
```suggestion
 * Add new `lint:yaml dirname --exclude=/dirname/foo.yaml --exclude=/dirname/bar.yaml`
```

christingruber replied:
👍🏻 good 👀

OskarStark commented on src/Symfony/Component/Yaml/Command/LintCommand.php:
```diff
@@ -76,6 +77,10 @@ protected function configure()
 
+You can also exclude one or more specific files:
+
+  <info>php %command.full_name% dirname --exclude="dirname/foo.yml" --exclude="dirname/bar.yml"</info>
```
```suggestion
  <info>php %command.full_name% dirname --exclude="dirname/foo.yaml" --exclude="dirname/bar.yaml"</info>
```

ogizanagi commented on src/Symfony/Component/Yaml/composer.json:
```diff
@@ -23,10 +23,10 @@
     },
     "require-dev": {
-        "symfony/console": "^4.4|^5.0|^6.0"
+        "symfony/console": "^5.3"
```
`^6.0` should still be allowed, right?

```suggestion
        "symfony/console": "^5.3|^6.0"
```

chalasr commented:
Thank you @christingruber.
